src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp

```diff
@@ -87,29 +87,195 @@ void ShenandoahDirectCardMarkRememberedSet::merge_overreach(size_t first_cluster
+    ShenandoahParallelWorkerSession worker_session(worker_id);
+    do_work(worker_id);
+  }
+}
+
+void ShenandoahScanRememberedTask::do_work(uint worker_id) {
   ShenandoahWorkerTimingsTracker x(ShenandoahPhaseTimings::init_scan_rset, ShenandoahPhaseTimings::ScanClusters, worker_id);
 
   ShenandoahObjToScanQueue* q = _queue_set->queue(worker_id);
   ShenandoahObjToScanQueue* old = _old_queue_set == NULL ? NULL : _old_queue_set->queue(worker_id);
   ShenandoahMarkRefsClosure<YOUNG> cl(q, _rp, old);
-  RememberedScanner* scanner = ShenandoahHeap::heap()->card_scan();
+  ShenandoahHeap* heap = ShenandoahHeap::heap();
+  RememberedScanner* scanner = heap->card_scan();
 
   // set up thread local closure for shen ref processor
   _rp->set_mark_closure(worker_id, &cl);
-  ShenandoahHeapRegion* region = _regions->next();
-  while (region != NULL) {
-    log_debug(gc)("ShenandoahScanRememberedTask::work(%u), looking at region " SIZE_FORMAT, worker_id, region->index());
+  struct ShenandoahRegionChunk assignment;
+  bool has_work = _work_list->next(&assignment);
+  while (has_work) {
+#ifdef ENABLE_REMEMBERED_SET_CANCELLATION
+    // This check is currently disabled to avoid crashes that occur
+    // when we try to cancel remembered set scanning
+    if (heap->check_cancelled_gc_and_yield(_is_concurrent)) {
+      return;
+    }
+#endif
+    ShenandoahHeapRegion* region = assignment._r;
+    log_debug(gc)("ShenandoahScanRememberedTask::do_work(%u), processing slice of region "
+                  SIZE_FORMAT " at offset " SIZE_FORMAT ", size: " SIZE_FORMAT,
+                  worker_id, region->index(), assignment._chunk_offset, assignment._chunk_size);
     if (region->affiliation() == OLD_GENERATION) {
-      scanner->process_region(region, &cl);
+      size_t cluster_size =
+        CardTable::card_size_in_words() * ShenandoahCardCluster<ShenandoahDirectCardMarkRememberedSet>::CardsPerCluster;
+      size_t clusters = assignment._chunk_size / cluster_size;
+      assert(clusters * cluster_size == assignment._chunk_size, "Chunk assignments must align on cluster boundaries");
+      HeapWord* end_of_range = region->bottom() + assignment._chunk_offset + assignment._chunk_size;
+
+      // During concurrent mark, region->top() equals TAMS with respect to the current young-gen pass.  */
+      if (end_of_range > region->top()) {
+        end_of_range = region->top();
+      }
+      scanner->process_region_slice(region, assignment._chunk_offset, clusters, end_of_range, &cl, false, _is_concurrent);
     }
-    region = _regions->next();
+    has_work = _work_list->next(&assignment);
   }
 }
+
+
+size_t ShenandoahRegionChunkIterator::calc_group_size() {
+  // First group does roughly half of heap, one region at a time.
+  // Second group does roughly one quarter of heap, half of a region at a time, and so on.
+  // Last group does the remnant of heap, one _smallest_chunk_size at a time.
+  // Round down.
+  return _heap->num_regions() / 2;
```

Comment:
I've rewritten the comment here in an attempt to make the intention more clear.

Reply:
I get it. I think it would be more clear if this method were called `initial_group_size` as the reduction of subsequent group sizes does not happen in this method.